### PR TITLE
[IPython] [opengl] Make OpenGL kernel print work in IPython & IDLE

### PR DESCRIPTION
**Pull request by archibate**
Related issue = depends #1222

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: python/taichi/lang/kernel.py [hunk 1]
```diff
@@ -456,7 +456,9 @@ def __call__(self, *args, **kwargs):
         instance_id, arg_features = self.mapper.lookup(args)
         key = (self.func, instance_id)
         self.materialize(key=key, args=args, arg_features=arg_features)
-        return self.compiled_functions[key](*args)
+        ret = self.compiled_functions[key](*args)
+        print(taichi_lang_core.pop_python_print_buffer(), end='')
```
Comment (k-ye) [inline python/taichi/lang/kernel.py:460]: I don't think this works for all platforms, does it? See https://github.com/taichi-dev/taichi/blob/37976e750e1579ff9328f3d25f1a63cb3b3292e2/tests/python/test_print.py#L17-L19. According to that comment, CUDA's print buffer isn't flushed until `ti.sync()` is called. Maybe we need to fix the CUDA backend first?

File: python/taichi/misc/image.py [hunk 2]
```diff
@@ -23,7 +23,30 @@ def imwrite(img, filename):
         comp = img.shape[2]
     assert comp in [1, 3, 4], "Image must be either RGB/RGBA or greyscale"
 
-    img = np.ascontiguousarray(img.swapaxes(0, 1)[::-1, :])
+    return img.swapaxes(0, 1)[::-1, :]
+
+
+
+def imdisplay(img):
```
Comment (k-ye) [inline python/taichi/misc/image.py:30]: Unrelated?

File: python/taichi/lang/kernel.py (hunk 1 above)
Comment (archibate) [reply python/taichi/lang/kernel.py:460]: How about this: move this line to ti.sync() and doc this behavior, so that we get the same behavior along backends.

Comment (k-ye) [reply python/taichi/lang/kernel.py:460]: Yep SGTM. (At one point I think `ti.sync` should only do one thing, but it also feels reasonable to flush the print buffer at synchronization...)

**Comment by archibate** [comment]
Note that this only fix for the OpenGL backend. Please unify the CPU+CUDA (@yuanming-hu) and Metal (@k-ye) behavior to `PythonPrintBuffer` later.

File: docs/syntax.rst [hunk 3]
```diff
@@ -254,6 +252,10 @@ Debug your program with ``print()`` in Taichi-scope. For example:
         before kernel
         after kernel
         inside kernel
-        after
+        after sync
+
+    This is due to the fact that GPU memories are only copied on demand, i.e.
+    only when the output data is accessed, or performance will be harmed.
```
Comment (yuanming-hu) [inline docs/syntax.rst:258]:
```suggestion
    This is because GPU memory is only copied when necessary, for performance considerations.
```

**Comment by archibate** [comment]
> Cool!
> 
> I think it makes sense to always call `ti.sync` after kernels with `print`. When people are printing (e.g. for debugging purposes), they usually don't really care about performance.

But how do I hook kernel invocations? According to your comments in `Kernel.__call__`, I'm afraid detecting `if self.has_print` will harm non-print kernel's performance.

**Comment by yuanming-hu** [comment]
> > Cool!
> > I think it makes sense to always call `ti.sync` after kernels with `print`. When people are printing (e.g. for debugging purposes), they usually don't really care about performance.
> 
> But how do I hook kernel invocations? According to your comments in `Kernel.__call__`, I'm afraid detecting `if self.has_print` will harm non-print kernel's performance.

We can do that in the C++ part.

**Comment by archibate** [comment]
> > > Cool!
> > > I think it makes sense to always call `ti.sync` after kernels with `print`. When people are printing (e.g. for debugging purposes), they usually don't really care about performance.
> > 
> > 
> > But how do I hook kernel invocations? According to your comments in `Kernel.__call__`, I'm afraid detecting `if self.has_print` will harm non-print kernel's performance.
> 
> We can do that in the C++ part.

No, we have to perform a Python `print` to make IDLE/IPython get the string to their GUI, instead of console stdout.
That's why I need the `print(ti.core.pop_print_buffer())` to pop C++ prints into Python prints.
